smithy-aws-traits/src/main/java/software/amazon/smithy/aws/traits/tagging/AwsTagIndex.java

```diff
@@ -0,0 +1,159 @@
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.utils.SetUtils;
+
+/**
+ * Index of AWS tagging trait information in a service closure.
+ */
+public final class AwsTagIndex implements KnowledgeIndex {
+    private final Set<ShapeId> tagEnabledServices = new HashSet<>();
+    private final Set<ShapeId> servicesWithAllTagOperations = new HashSet<>();
+    private final Map<ShapeId, Set<ShapeId>> serviceToTaggableResources = new HashMap<>();
+    private final Map<ShapeId, Boolean> resourceIsTagOnCreate = new HashMap<>();
+    private final Map<ShapeId, Boolean> resourceIsTagOnUpdate = new HashMap<>();
+
+    private AwsTagIndex(Model model) {
+        PropertyBindingIndex propertyBindingIndex = PropertyBindingIndex.of(model);
+        OperationIndex operationIndex = OperationIndex.of(model);
+
+        for (ServiceShape service : model.getServiceShapesWithTrait(TagEnabledTrait.class)) {
+            tagEnabledServices.add(service.getId());
+            for (ShapeId resourceId : service.getResources()) {
+                ResourceShape resource = model.expectShape(resourceId)
+                        .asResourceShape().get();
+                Set<ShapeId> taggableResources = new HashSet<>();
+                if (resource.hasTrait(TaggableTrait.class)) {
+                    taggableResources.add(resourceId);
+
+                    // Check if tag property is specified on create.
+                    resourceIsTagOnCreate.put(resourceId, computeTagOnCreate(model, resource, propertyBindingIndex));
+                    // Check if tag property is specified on update.
+                    resourceIsTagOnUpdate.put(resourceId, computeTagOnUpdate(model, resource, propertyBindingIndex));
+                }
+                serviceToTaggableResources.put(service.toShapeId(),
+                        SetUtils.copyOf(taggableResources));
+            }
+
+            //Check if service has the three service-wide tagging operations unbound to any resource.
+            if (verifyTagApis(model, service, operationIndex)) {
+                servicesWithAllTagOperations.add(service.getId());
+            }
+        }
+    }
+
+    private boolean verifyTagApis(Model model, ServiceShape service, OperationIndex operationIndex) {
+        boolean hasTagApi = false;
+        boolean hasUntagApi = false;
+        boolean hasListTagsApi = false;
+
+        ShapeId tagResourceId = ShapeId.fromParts(service.getId().getNamespace(),
+                                    TaggingShapeUtils.TAG_RESOURCE_OPNAME);
+            if (service.getOperations().contains(tagResourceId)) {
+                OperationShape tagResourceOperation = model.expectShape(tagResourceId).asOperationShape().get();
+                Map<String, MemberShape> inputMembers = operationIndex.getInputMembers(tagResourceOperation);
+
+                hasTagApi = inputMembers.entrySet().stream().filter(memberEntry ->
```

Comment:
I disagree that the code blocks are duplicated. The logic all three share exactly is whether or not resource ARN is in the input. TagResource iterates over input members looking for a tag list or key/value pairs member. UntagResource iterates over input members looking for a tag keys member. ListTagsForResoruce iterates over output members looking for a tag list or key/value pairs. Trying to break down those pieces and re-use some of the generic iteration and logic would introduce unnecessary complexity and hard to read methods with low reuse. All that and I don't think the core logic would have been clearer, if at all.

I didn't want to use a for-loop because getting a boolean value seemed fairly clean through an expression rather than flow control -> collection storage ->  measure size of collection.